### PR TITLE
fix spurious disconnect due to block download crowding out getheaders

### PR DESCRIPTION
**Pull request by gandrewstone**
And convert all logs to GetLogName()

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -7040,9 +7043,9 @@ bool SendMessages(CNode *pto)
                 {
                     state.fSyncStarted = true;
                     state.nSyncStartTime = GetTime();
-                    state.fFirstHeadersReceived = false;
+                    // state.fFirstHeadersReceived = false;  // Only require headers within some time the first time
                     state.fRequestedInitialBlockAvailability = true;
```
Comment (ptschip) [inline src/main.cpp]: i think you can remove this commented out ilne

File: src/main.cpp [hunk 2]
```diff
@@ -6420,6 +6420,9 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
                 SendExpeditedBlock(*pblock, pfrom);
         }
 
+        CNodeState* state = State(pfrom->GetId());
+        if (state) state->nSyncStartTime = GetTime();  // reset the time because more headers needed
+
```
Comment (ptschip) [inline src/main.cpp]: state still requires cs_main lock, also you should put in a DbgAssert as we do for all other state lookups... so 
{
    LOCK(cs_main);
    CNodeState *state = State(nodeid);
    DbgAssert(state != nullptr, return );
    if (state) state->nSyncStartTime = GetTime();
}

File: src/main.cpp [hunk 3]
```diff
@@ -6106,7 +6106,8 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
             LOG(NET, "more getheaders (%d) to end to peer=%s (startheight:%d)\n", pindexLast->nHeight,
                 pfrom->GetLogName(), pfrom->nStartingHeight);
             pfrom->PushMessage(NetMsgType::GETHEADERS, chainActive.GetLocator(pindexLast), uint256());
-
+            CNodeState* state = State(pfrom->GetId());
+            if (state) state->nSyncStartTime = GetTime();  // reset the time because more headers needed
             // During the process of IBD we need to update block availability for every connected peer. To do that we
```
Comment (ptschip) [inline src/main.cpp]: nit -> leave a newline after this ...

Also need a dbgassert here

Comment (sickpig) [reply src/main.cpp]: plus the lock on `cs_main` like requested for line 6423?

File: src/main.cpp [hunk 4]
```diff
@@ -7040,9 +7043,9 @@ bool SendMessages(CNode *pto)
                 {
                     state.fSyncStarted = true;
                     state.nSyncStartTime = GetTime();
-                    state.fFirstHeadersReceived = false;
+                    // state.fFirstHeadersReceived = false;  // Only require headers within some time the first time
```
Comment (sickpig) [inline src/main.cpp]: I don't get the meaning of the comment.

File: src/main.cpp [hunk 5]
```diff
@@ -7040,9 +7043,9 @@ bool SendMessages(CNode *pto)
...
                     state.fRequestedInitialBlockAvailability = true;
-                    state.nFirstHeadersExpectedHeight = pindexBestHeader->nHeight;
+                    state.nFirstHeadersExpectedHeight = pindexBestHeader->nHeight-2;  // -2 to give room for slow sync
```
Comment (sickpig) [inline src/main.cpp]: What do you mean w/ "slow sync"? peers that are slow at providing or peers that are a little bit behind the actual net chaintip? 

if the latter is true I wonder if we could expand the comment a little.

**Review by sickpig — COMMENTED** [review]
nit: as a general point it would have being great to have the changes to the LOG statements isolated on their own commit so that the functional code changes could have been easier to identify and review.

**Comment by sickpig** [comment]
travis failed due to formatting issue.

File: src/main.cpp (hunk 3 above)
Comment (ptschip) [reply src/main.cpp]: cs_main is already locked further up in this instance.

Comment (sickpig) [reply src/main.cpp]: fair enough

File: src/main.cpp [hunk 6]
```diff
@@ -6106,8 +6102,12 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
             LOG(NET, "more getheaders (%d) to end to peer=%s (startheight:%d)\n", pindexLast->nHeight,
                 pfrom->GetLogName(), pfrom->nStartingHeight);
             pfrom->PushMessage(NetMsgType::GETHEADERS, chainActive.GetLocator(pindexLast), uint256());
-            CNodeState* state = State(pfrom->GetId());
-            if (state) state->nSyncStartTime = GetTime();  // reset the time because more headers needed
+
+            CNodeState *state = State(pfrom->GetId());
+            DbgAssert(state != nullptr, );
+            if (state)
```
Comment (ptschip) [inline src/main.cpp]: i think it needs a return false if dbgassert

File: src/main.cpp [hunk 7]
```diff
@@ -6420,8 +6420,14 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
                 SendExpeditedBlock(*pblock, pfrom);
         }
 
-        CNodeState* state = State(pfrom->GetId());
-        if (state) state->nSyncStartTime = GetTime();  // reset the time because more headers needed
+        {
+            LOCK(cs_main);
+            CNodeState *state = State(pfrom->GetId());
+            DbgAssert(state != nullptr, );
+            if (state)
```
Comment (ptschip) [inline src/main.cpp:6436]: same here, return false

**Review by ptschip — APPROVED** [review]

**Review by ptschip — APPROVED** [review]

**Pull request merged**